pkg/workloadlearning/cache.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2025 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workloadlearning
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"github.com/pingcap/tidb/pkg/kv"
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"github.com/pingcap/tidb/pkg/util"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"go.uber.org/zap"
+)
+
+// ReadTableCostCache stores the cached workload learning metrics
+type ReadTableCostCache struct {
+	TableCostMetrics map[int64]*ReadTableCostMetrics // key: tableID
+	Version          uint64
+}
+
+// WLCacheWorker the worker to cache all workload-related metrics
+// Now it is also used to save the cache data of table cost metrics.
+type WLCacheWorker struct {
+	sysSessionPool     util.DestroyableSessionPool
+	readTableCostCache *ReadTableCostCache
+	sync.RWMutex
+}
+
+// NewWLCacheWorker Create a new workload learning cache worker to cache all workload-related metrics
+// from storage mysql.tidb_workload_values to memory
+func NewWLCacheWorker(pool util.DestroyableSessionPool) *WLCacheWorker {
+	return &WLCacheWorker{pool, &ReadTableCostCache{}, sync.RWMutex{}}
+}
+
+// UpdateTableCostCache refreshes the cached workload learning metrics
+func (cw *WLCacheWorker) UpdateTableCostCache() {
+	// Get latest metrics from storage
+	se, err := cw.sysSessionPool.Get()
+	if err != nil {
+		logutil.BgLogger().Warn("Get system session failed when updating table cost cache", zap.Error(err))
+		return
+	}
+	defer func() {
+		if err == nil { // only recycle when no error
+			cw.sysSessionPool.Put(se)
+		} else if err != nil && se != nil {
+			// Note: Otherwise, the session will be leaked.
+			cw.sysSessionPool.Destroy(se)
+		}
+		// If err != nil && session is nil, which means the session pool is closed, no need to recycle and destroy
+	}()
+
+	sctx := se.(sessionctx.Context)
+	exec := sctx.GetRestrictedSQLExecutor()
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnWorkloadLearning)
+
+	// Whether to update table cost metrics
+	// Get the latest latestVersionInStorage in the storage
+	sql := `SELECT version FROM mysql.tidb_workload_values
+            WHERE category = %? AND type = %?
+            ORDER BY version DESC LIMIT 1`
+	rows, _, err := exec.ExecRestrictedSQL(ctx, nil, sql, feedbackCategory, tableCostType)
+	if err != nil {
+		logutil.BgLogger().Warn("Failed to get the latest table cost version", zap.Error(err))
```

Comment:
Do you want to print the error stack here? You might need to use ErrVerboseLogger.

Reply:
Changed